core/templates/services/messenger.service.ts

```diff
@@ -19,147 +19,196 @@
-      },
-      explorationReset: function(payload) {
-        return Boolean(payload.stateName);
-      },
-      explorationCompleted: function() {
-        return true;
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+
+import { LoggerService } from 'services/contextual/logger.service';
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+interface IHeightChangeData {
+  height: number;
+  scroll: number;
+}
+
+interface IExplorationLoadedData {
+  explorationVersion: number;
+  explorationTitle: string;
+}
+
+interface IStateTransitionData {
+  explorationVersion: number;
+  oldStateName: string;
+  jsonAnswer: string;
+  newStateName: string;
+}
+
+interface IExplorationCompletedData {
+  explorationVersion: number;
+}
+
+interface IMessageValidatorsType {
+  heightChange: ((payload: IHeightChangeData) => boolean);
+  explorationLoaded: (() => boolean);
+  stateTransition: (
+    (payload: IStateTransitionData) => boolean);
+  explorationReset: ((payload: {stateName: string}) => boolean);
+  explorationCompleted: (() => boolean);
+}
+
+interface GetPayloadType {
+  heightChange: ((data: IHeightChangeData) => IHeightChangeData);
+  explorationLoaded: ((data: IExplorationLoadedData) => IExplorationLoadedData);
+  stateTransition: ((data: IStateTransitionData) => IStateTransitionData);
+  explorationCompleted: (
+    (data: IExplorationCompletedData) => IExplorationCompletedData);
+  explorationReset: ((data: string) => {stateName: string});
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class MessengerService {
+  constructor(
+    private loggerService: LoggerService, private windowRef: WindowRef) {}
+
+  // TODO(brianrodri): Move these into a .constants.ts file.
+  HEIGHT_CHANGE: string = 'heightChange';
+  EXPLORATION_LOADED: string = 'explorationLoaded';
+  STATE_TRANSITION: string = 'stateTransition';
+  EXPLORATION_RESET: string = 'explorationReset';
+  EXPLORATION_COMPLETED: string = 'explorationCompleted';
+
+  SUPPORTED_HASHDICT_VERSIONS: Set<string> = (
+    new Set(['0.0.0', '0.0.1', '0.0.2', '0.0.3']));
+
+  MESSAGE_VALIDATORS: IMessageValidatorsType = {
+    heightChange: (payload: IHeightChangeData): boolean => {
+      const {height, scroll} = payload;
+      return (
+        Number.isInteger(height) && height > 0 && typeof scroll === 'boolean');
+    },
+    explorationLoaded(): boolean {
+      return true;
+    },
+    stateTransition(payload: IStateTransitionData): boolean {
+      return Boolean(payload.oldStateName) || Boolean(payload.newStateName);
+    },
+    explorationReset(payload: {stateName: string}): boolean {
+      return Boolean(payload.stateName);
+    },
+    explorationCompleted(): boolean {
+      return true;
+    }
+  };
+
+  getPayload: GetPayloadType = {
+    heightChange(data: IHeightChangeData) {
```

Comment:
Shouldn't we have return type of this function (and below)?

Reply:
Updated.